src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -0,0 +1,682 @@
+
+  /// \brief Name of the model to which the sensor is attached.
+  public: std::string modelName;
+
+  /// \brief Message for visualizing contact positions
+  public: ignition::msgs::Marker positionMarkerMsg;
+
+  /// \brief Radius of the visualized contact sphere
+  public: double contactRadius{0.10};
+
+  /// \brief Message for visualizing contact forces
+  public: ignition::msgs::Marker forceMarkerMsg;
+
+  /// \brief Length of the visualized force cylinder
+  public: double forceLenght{0.20};
+
+  /// \brief Position interpolated from the Contact messages
+  public: std::vector<ignition::math::Vector3d> interpolatedPosition;
+
+  /// \brief Pose of the sensor model
+  public: ignition::math::Pose3f sensorWorldPose;
+
+  /// \brief Whether a new message has been returned by the Depth Camera
+  public: bool newCameraMsg{false};
+
+  /// \brief Depth Camera update rate
+  public: float cameraUpdateRate{1};
+
+  /// \brief Message returned by the Depth Camera
+  public: ignition::msgs::PointCloudPacked cameraMsg;
+
+  /// \brief 3D vector to store unpacked XYZ points
+  std::vector<std::vector<std::vector<float>>> imageXYZ;
+};
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::Load(const EntityComponentManager &_ecm,
+                                       const sdf::ElementPtr &_sdf)
+{
+    igndbg << "Loading plugin [OpticalTactilePlugin]" << std::endl;
+
+    // Get sdf parameters
+    if (!_sdf->HasElement("resolution"))
+    {
+        ignerr << "Missing required parameter <resolution>." << std::endl;
+        return;
+    }
+    else
+    {
+        this->resolution = _sdf->Get<int>("resolution");
+    }
+
+    if (!_sdf->HasElement("visualize_forces"))
+    {
+        ignerr << "Missing required parameter <visualize_forces>." << std::endl;
+        return;
+    }
+    else
+    {
+        this->visualize_forces = _sdf->Get<bool>("visualize_forces");
+    }
+
+    // Get all the sensors
+    auto allLinks =
+        _ecm.ChildrenByComponents(this->model.Entity(), components::Link());
+
+    for (const Entity linkEntity : allLinks)
+    {
+        auto linkCollisions =
+            _ecm.ChildrenByComponents(linkEntity, components::Collision());
+        for (const Entity colEntity : linkCollisions)
+        {
+            if (_ecm.EntityHasComponentType(colEntity,
+                components::ContactSensorData::typeId))
+            {
+                this->sensorEntities.push_back(colEntity);
+
+                this->modelName = this->model.Name(_ecm);
+
+                igndbg << "Sensor detected within model "
+                    << this->modelName << std::endl;
+            }
+        }
+    }
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::Enable(const bool _value)
+{
+    // Just a placeholder method for the moment
+    _value;
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePlugin::Update(const UpdateInfo &_info,
```

Comment:
We should document this somewhere, but in general, we don't recommend using `System::Update` call unless absolutely necessary. This is the phase where physics is running and the contact points may not be ready yet. Consider putting this in `PostUpdate`.

Reply:
Thanks! I was not aware of that.

Reply:
> We should document this somewhere,

There's a bit [here](https://ignitionrobotics.org/api/gazebo/3.0/createsystemplugins.html):

_Generally, systems that do not manage physics stepping will not need to implement the ISystemUpdate interface._

And the [System](https://ignitionrobotics.org/api/gazebo/3.0/classignition_1_1gazebo_1_1System.html) API docs also mention:

_Update (...) Used for physics simulation step_

But we can always make it clearer.

Reply:
I will admit we did find that page (I see it in our Slack records), but I didn't read it closely... oops.

Reply:
My bad, I was aware of that documentation as well. I would suggest to specify what exactly _physics simulation step_ is or writing an example, so it's easy to know for newcomers if they should implement it. I'm not sure what it is. 

Anyway, moving to `PostUpdate`. Thanks!